src/typ.rs

```diff
@@ -2271,6 +2271,21 @@ enum UnifyError {
     RecursiveType,
 }
 
+fn unify_enclosed_type(
+    e1: &Type,
+    e2: &Type,
+    result: Result<(), UnifyError>,
+) -> Result<(), UnifyError> {
+    // If types cannot unify, show the type error with the enclosing types, e1 and e2.
+    match result {
+        Err(_) => Err(UnifyError::CouldNotUnify {
```

Comment:
Could we match specifically on `Err(UnifyError::CouldNotUnify { .. })`? To be sure it's the right error we're overriding